src/arithmetic.cpp

```diff
@@ -1806,6 +1763,52 @@ void __mul_ex(fp* z, const fp* x, const fp* y)
     asm("pop %r15;");
     asm("pop %rbp;");
 }
+
+typedef void (*blsmul_func_t)(fp*, const fp*, const fp*);
+
+static bool cpu_has_bmi2_and_adx() {
+    // borrowed from: https://github.com/Mysticial/FeatureDetector/blob/master/src/x86/cpu_x86.cpp
+    int32_t info[4];
+    __cpuid_count(0, 0, info[0], info[1], info[2], info[3]);
+    int nIds = info[0];
+    if(nIds >= 0x00000007) {
+        __cpuid_count(0x00000007, 0, info[0], info[1], info[2], info[3]);
+        if((info[1] & (1 <<  8)) && (info[1] & (1 << 19))) // BMI2 && ADX
+            return true;
+    }
+    return false;
```

Comment:
Should we use instead:

```
   __builtin_cpu_init ();
   return __builtin_cpu_supports("bmi2");
```

see [this](https://gcc.gnu.org/onlinedocs/gcc/x86-Built-in-Functions.html#index-_005f_005fbuiltin_005fcpu_005fsupports-1).

Seems to work for both clang and gcc.

Reply:
It's not just BMI2 that's used, but also ADX. Even clang16 doesn't support `__builtin_cpu_supports("adx")`, just gcc11+

Of course, in practice, it may be that there are no CPUs out there that have BMI2 but not ADX. Hard to say though.. best be on the safe side I think